src/elm/expression.ts

```diff
@@ -46,6 +55,20 @@ export class Expression {
       return null;
     }
   }
+
+  /**
+   * Function used in error reporting during execution. Retrieves the source library from
+   * the context if it exists, or recursively traverses the context's parents until a source
+   * library identifier and version are found.
+   */
+  getRecursiveLibraryIdentifier(ctx: Context): string {
+    if (ctx instanceof PatientContext || ctx instanceof UnfilteredContext) {
```

Comment:
Makes sense. Only annoying thing about this is since `ctx` is of type `Context` (which does not have a `.library` property), TypeScript gets mad because `library` doesn't exist on `ctx`. I could cast `ctx` to `any` both in the if and in the actually code for accessing the identifier.

I do agree that it would be nice to write the `if` in a more maintainable way that doesn't scale linearly with each new context added. Let me know if the cast is sufficient

Reply:
Yeah, I think casting is sufficient. If you want to be fancy, you could create a throw-away interface like:
```ts
interface ContextWithOptionalLibrary extends Context {
  library?: Library;
}
```
and cast to that. Then you still get all the type-checking since it knows what `library` is.